### DISCUSSION TITLE
Problem with static site generation and dynamic rendering

### DISCUSSION THREAD

**Discussion by weilueluo** [vercel/next.js#50661] [App Router]
So currently, a route opts into dynamic rendering if it uses features like `headers` because the server will not know these values during build time.

However, the developers may utilize the headers to render only a countable number of pages, and it makes sense to allow developers to enumerate these pages and provide the ability to generate them during build time, similar to `generateStaticParams`.

An example of this would be rendering themed pages based on cookies in the headers, let's say there are only 2 possible values for the theme: `light` and `dark`, I would like the server to serve different versions of the HTML based on the cookies I left in their browser last time they visited. At the same time being able to use features like `generateStaticParams` in the sub-routes. Ideally, I would like a way to provide these `light` and `dark` values during build time, but there is currently no elegant way to do this unless we put the theme as a path variable which is ugly.

Generally, I would like to have the option to provide enumerated values for dynamic features so that I can utilize static site generation.  How feasible is this approach? Any idea?

*New to NextJs app router, I would very much appreciate it if you could correct any misunderstanding I have*

**Comment by ben519** [comment]
Did you find a solution to this? I see you marked it as closed.

**Reply by weilueluo** [reply]
Hey Ben, I did not find a solution to this, instead, I think I just misunderstood what `generateStaticParams` was offering. I think `generateStaticParams`'s main use case is to provide a way to output a static site, like what we have for `next export`, in this case, there is no way for the server to handle dynamic features because there is simply no server at the first place --- we are just serving static HTML files. 

When we use dynamic features, although we do not have the option to generate them statically, at the end of the day, these pages will still be generated on the server, we still get all the benefits of server-side rendering --- the only difference is when the page gets render: during build time or during request time. This means we will only affect the first user who visits the site.

I believe NextJs can still provide the option to pre-render some of the dynamic pages, but I feel this is no longer an issue for my use case, but since you have asked, I will reopen the discussion, so that maybe someone who is more expert in NextJs can correct my words or provide a solution.

In the meantime, I think if you really want to emulate the behaviour of `generateStaticParams` without having the variable in the path, we will just need to export the `revalidate` variable from the `page.js` and set it to an arbitrarily large number.
